examples/akka-persistence-app/src/main/scala/org.virtuslab/example/ShoppingCart.scala

```diff
@@ -0,0 +1,123 @@
+import akka.persistence.typed.scaladsl.Effect
+import akka.persistence.typed.scaladsl.EventSourcedBehavior
+import akka.persistence.typed.scaladsl.ReplyEffect
+import akka.persistence.typed.scaladsl.RetentionCriteria
+import io.circe.Codec
+import io.circe.generic.semiauto.deriveCodec
+
+import org.virtuslab.ash.circe.AkkaCodecs
+
+object ShoppingCart {
+
+  val EntityKey: EntityTypeKey[Command] =
+    EntityTypeKey[Command]("ShoppingCart")
+
+  def init(system: ActorSystem[_]): Unit = {
+    ClusterSharding(system).init(Entity(EntityKey) { entityContext =>
+      ShoppingCart(entityContext.entityId)
+    })
+  }
+
+  def apply(cartId: String): Behavior[Command] = {
+    EventSourcedBehavior
+      .withEnforcedReplies[Command, Event, State](
+        persistenceId = PersistenceId(EntityKey.name, cartId),
+        emptyState = State.empty,
+        commandHandler = (state, command) => handleCommand(cartId, state, command),
+        eventHandler = (state, event) => handleEvent(state, event))
+      .withRetention(RetentionCriteria.snapshotEvery(numberOfEvents = 100, keepNSnapshots = 3))
+      .onPersistFailure(SupervisorStrategy.restartWithBackoff(200.millis, 5.seconds, 0.1))
+  }
+
+  /**
+   * This interface defines all the commands (messages) that the ShoppingCart actor supports.
+   */
+  sealed trait Command extends CirceAkkaSerializable
+
+  /**
+   * A command to add an item to the cart.
+   *
+   * It replies with `StatusReply[Summary]`, which is sent back to the caller when all the events emitted by this
+   * command are successfully persisted.
+   */
+  final case class AddItem(itemId: String, quantity: Int, replyTo: ActorRef[StatusReply[Summary]]) extends Command
+
+  /**
+   * Summary of the shopping cart state, used in reply messages.
+   */
+  final case class Summary(items: Map[String, Int]) extends CirceAkkaSerializable
+
+  /**
+   * This interface defines all the events that the ShoppingCart supports.
+   */
+  sealed trait Event extends CirceAkkaSerializable {
+    def cartId: String
+  }
+
+  final case class ItemAdded(cartId: String, itemId: String, quantity: Int) extends Event
+
+  /**
+   * State of the shopping cart
+   * @param items
+   *   \- Map containing information about items that are in the cart currently
+   */
+  final case class State(items: Map[String, Int]) extends CirceAkkaSerializable {
+
+    def hasItem(itemId: String): Boolean =
+      items.contains(itemId)
+
+    def isEmpty: Boolean =
+      items.isEmpty
+
+    def updateItem(itemId: String, quantity: Int): State = {
+      quantity match {
+        case 0 => copy(items = items - itemId)
+        case _ => copy(items = items + (itemId -> quantity))
+      }
+    }
+  }
+  object State {
+    val empty: State = State(items = Map.empty)
+  }
+
+  // codecs derivation for all types extending CirceAkkaSerializable
+  implicit lazy val codecActorRefResponse: Codec[ActorRef[StatusReply[Summary]]] = new AkkaCodecs {}.actorRefCodec
```

Comment:
Huh while you're here... maybe you could add `object AkkaCodecs extends AkkaCodecs`? coz `new ... {}` looks awkward as hell